XMonad/Util/NamedScratchpad.hs

```diff
@@ -152,6 +154,31 @@ allNamedScratchpadAction :: NamedScratchpads
                          -> X ()
 allNamedScratchpadAction = someNamedScratchpadAction mapM_ runApplication
 
+-- | A @logHook@ to hide scratchpads when they lose focus.  This can be
+-- useful for e.g. dropdown terminals.  Note that this also requires you
+-- to use the 'XMonad.Hooks.RefocusLast.refocusLastLogHook'.
+--
+-- ==== __Example__
+--
+-- > import XMonad.Hooks.RefocusLast (refocusLastLogHook)
+-- > import XMonad.Util.NamedScratchpad
+-- >
+-- > main = xmonad $ def
+-- >   { logHook = refocusLastLogHook
+-- >            >> nsHideOnFocusLoss myScratchpads
+-- >               -- enable hiding for all of @myScratchpads@
+-- >   }
+nsHideOnFocusLoss :: NamedScratchpads -> X ()
+nsHideOnFocusLoss scratches = withWindowSet $ \winSet -> do
+    let cur = W.currentTag winSet
+    withRecentsIn cur () $ \lastFocus _ ->
+        when (lastFocus `elem` W.index winSet && cur /= scratchpadWorkspaceTag) $
+            whenX (isNS lastFocus) $
+                shiftToNSP (W.workspaces winSet) ($ lastFocus)
+  where
+    isNS :: Window -> X Bool
+    isNS w = or <$> traverse ((`runQuery` w) . query) scratches
```

Comment:
Hmm, can this be useful somewhere else? It looks handy

Reply:
There's also https://hackage.haskell.org/package/xmonad-contrib-0.16/docs/src/XMonad.Util.Loggers.NamedScratchpad.html#isSp, but it's not quite the same thing and arguably misnamed as it does more than just identifying named scratchpads.

Reply:
I mean, we _could_ make this a top-level function (I haven't actually checked if this could be useful anywhere else, but I imagine we'd find places).  Then, however I propose we go the extra mile and rewrite this as a proper 

``` haskell
  isNS scratches w = anyM ((`runQuery` w) . query) scratches
``` 

for a proper definition of `anyM`.  Reason being that `or <$> traverse` (as opposed to `or $ map`) does not actually short-circuit when a match has been found (yay for applicative effects).  There are "correct" definitions of `orM` and `anyM` that do this, but they have to be defined via some custom fold and are thus more than one line.

Reply:
Interesting. You don't have to delay merging this though, we can take care of it later